R/sim_linelist.R

```diff
@@ -0,0 +1,217 @@
+#' `"confirmed"`. Values of each case type must sum to one.
+#' @param config A list of settings to adjust the randomly sampled delays and
+#' Ct values (if `add_ct = TRUE`). See [create_config()] for more information.
+#' @param ... [dots] Extra arguments to be passed to other functions.
+#' **Currently not used and will return a warning if extra arguments are
+#' supplied**.
+#'
+#' @return A line list `<data.frame>`
+#' @export
+#'
+#' @author Joshua W. Lambert, Carmen Tamayo
+#'
+#' @examples
+#' # load data required to simulate line list
+#' serial_interval <- epiparameter::epidist(
+#'   disease = "COVID-19",
+#'   epi_dist = "serial interval",
+#'   prob_distribution = "gamma",
+#'   prob_distribution_params = c(shape = 1, scale = 1)
+#' )
+#'
+#' # get onset to hospital admission from {epiparameter} database
+#' onset_to_hosp <- epiparameter::epidist_db(
+#'   disease = "COVID-19",
+#'   epi_dist = "onset to hospitalisation",
+#'   single_epidist = TRUE
+#' )
+#'
+#' # get onset to death from {epiparameter} database
+#' onset_to_death <- epiparameter::epidist_db(
+#'   disease = "COVID-19",
+#'   epi_dist = "onset to death",
+#'   single_epidist = TRUE
+#' )
+#' # example with single hospitalisation rate for entire population
+#' linelist <- sim_linelist(
+#'   R = 1.1,
+#'   serial_interval = serial_interval,
+#'   onset_to_hosp = onset_to_hosp,
+#'   onset_to_death = onset_to_death,
+#'   hosp_rate = 0.5
+#' )
+#'
+#' # example with age-stratified hospitalisation rate
+#' # 20% for over 80s
+#' # 10% for under 5s
+#' # 5% for the rest
+#' age_dep_hosp_rate <- data.frame(
+#'   age_limit = c(1, 5, 80),
+#'   rate = c(0.1, 0.05, 0.2)
+#' )
+#' linelist <- sim_linelist(
+#'   R = 1.1,
+#'   serial_interval = serial_interval,
+#'   onset_to_hosp = onset_to_hosp,
+#'   onset_to_death = onset_to_death,
+#'   hosp_rate = age_dep_hosp_rate
+#' )
+sim_linelist <- function(R,
+                         serial_interval,
+                         onset_to_hosp,
+                         onset_to_death,
+                         hosp_rate = 0.2,
+                         hosp_death_rate = 0.5,
+                         non_hosp_death_rate = 0.05,
+                         outbreak_start_date = as.Date("2023-01-01"),
+                         add_names = TRUE,
+                         add_ct = FALSE,
+                         min_chain_size = 10,
+                         population_age = c(1, 90),
+                         case_type_probs = c(
+                           suspected = 0.2,
+                           probable = 0.3,
+                           confirmed = 0.5
+                         ),
+                         config = create_config(),
+                         ...) {
+  chkDots(...)
+
+  serial_interval <- as.function(serial_interval, func_type = "generate")
+  onset_to_hosp <- as.function(onset_to_hosp, func_type = "generate")
+  onset_to_death <- as.function(onset_to_death, func_type = "generate")
```

Comment:
I'm still slightly uneasy with this.

Functions and epidist objects are very different and passing them as if they are the same feels strange. I'm afraid this could teach users that they can use any objects they like in any functions and it will magically work and this seems like a disservice that will result in a lot of frustration.

An alternative could be "a function or something coercible to one via `as.function()`". The problem is that we use the extra `func_type` argument which will likely not exist in other methods.

On the note of foreign methods, there is also a risk for an uncontrolled behaviour here if an object with an `as.function()` method which doesn't behave as what you expect is passed.

Reply:
I believe that the current implementation benefits for allowing the base R approach of directly supplying a function, while also enhancing this by allowing `<epidist>` objects to be directly passed to the function, which is a feature other people have mentioned they preferred in previous versions of other Epiverse packages. 

In 13d9ee1431d581687e83146f5b504ac76d2ca230 I have improved the input checking as I agree the uncontrolled behaviour is unwanted, but it can be avoided with defensive programming and good documentation. 

I disagree in documenting the functionality around `as.function()` as this could include many classes that will not be compatible with our functions. Instead having it clearly communicated that the default (or dependency-less) option be the anonymous function, with the added enhancement provided by {epiparameter} and utilised by {simulist}.